compiler/z/codegen/OMRCodeGenerator.cpp

```diff
@@ -2609,21 +2609,59 @@ OMR::Z::CodeGenerator::prepareRegistersForAssignment()
          }
       }
 
+   TR::Machine* machine = self()->machine();
+
    // Lock vector registers. This is done for testing purposes and to artificially increase register pressure
    // We go from the end since in most cases, we want to test functionality of the first half i.e 0-15 for overlap with FPR etc
-   if (TR::Options::_numVecRegsToLock > 0  &&
-       TR::Options::_numVecRegsToLock <= (TR::RealRegister::LastAssignableVRF - TR::RealRegister::FirstAssignableVRF))
+   if (TR::Options::_numVecRegsToLock > 0  && TR::Options::_numVecRegsToLock <= (TR::RealRegister::LastAssignableVRF - TR::RealRegister::FirstAssignableVRF))
+       {
+       for (int32_t i = TR::RealRegister::LastAssignableVRF; i > (TR::RealRegister::LastAssignableVRF - TR::Options::_numVecRegsToLock); --i)
+          {
+          machine->getRegisterFile(i)->setState(TR::RealRegister::Locked);
+          }
+       }
+
+   int32_t lockedGPRs = 0;
+   int32_t lockedHPRs = 0;
+   int32_t lockedFPRs = 0;
+   int32_t lockedVRFs = 0;
+
+   // count up how many registers are locked for each type
+   for (int32_t i = TR::RealRegister::FirstGPR; i <= TR::RealRegister::LastGPR; ++i)
       {
-      for(int32_t i = TR::RealRegister::LastAssignableVRF;
-              i > (TR::RealRegister::LastAssignableVRF - TR::Options::_numVecRegsToLock);
-              i--)
-         {
+      TR::RealRegister* realReg = machine->getS390RealRegister((TR::RealRegister::RegNum)i);
+      if (realReg->getState() == TR::RealRegister::Locked)
+         ++lockedGPRs;
+      }
 
-         self()->machine()->getRegisterFile(i)->setState(TR::RealRegister::Locked);
-         }
+   for (int32_t i = TR::RealRegister::FirstHPR; i <= TR::RealRegister::LastHPR; ++i)
+      {
+      TR::RealRegister* realReg = machine->getS390RealRegister((TR::RealRegister::RegNum)i);
+      if (realReg->getState() == TR::RealRegister::Locked)
+         ++lockedHPRs;
+      }
+
+   for (int32_t i = TR::RealRegister::FirstFPR; i <= TR::RealRegister::LastFPR; ++i)
+      {
+      TR::RealRegister* realReg = machine->getS390RealRegister((TR::RealRegister::RegNum)i);
+      if (realReg->getState() == TR::RealRegister::Locked)
+         ++lockedFPRs;
       }
+
+   for (int32_t i = TR::RealRegister::FirstVRF; i <= TR::RealRegister::LastVRF; ++i)
+      {
+      TR::RealRegister* realReg = machine->getS390RealRegister((TR::RealRegister::RegNum)i);
+      if (realReg->getState() == TR::RealRegister::Locked)
+         ++lockedVRFs;
+      }
+
+   machine->setNumberOfLockedRegisters(TR_GPR, lockedGPRs);
+   machine->setNumberOfLockedRegisters(TR_HPR, lockedHPRs);
+   machine->setNumberOfLockedRegisters(TR_FPR, lockedFPRs);
+   machine->setNumberOfLockedRegisters(TR_VRF, lockedVRFs);
+
```

Comment:
I think I realized at some point in the past that we really don't need to count the number of locked regs because they are locked from the outset of the compilation and never change. It was a minor issue that was never really revisited, but since you're in this part of the code base doing good things can you look into it? The locked regs are determined by the linkage; each linkage sets the register file up in the state that it wants, so that might be a good place to count the number of locked regs as well.

Reply:
@ymanton indeed this is how it is on Power. We simply initialize with a constant. Unfortunately on Z it is a little bit more complicated since there are many places where we lock register depending on the state of the JIT and whether some optimizations ran [1] [2] [3].

With this in mind the most sane thing to do is just do a single pass through the real registers and count them up for each compilation. I agree this is not the most efficient thing to do, but the alternative would be to keep incrementing and decrementing the counts via `getNumberOfLockedRegisters` / `setNumberOfLockedRegisters` each time we change a locked state.

I will be simplifying the logic in this area in following PRs so hopefully we can get it to a manageable state so we can do exactly what Power does and simply initialize the number of locked registers with a constant. This is definitely in my crosshairs.

[1] https://github.com/eclipse/omr/blob/fcfd166b487d3796b525be3def5afe8b97c2c022/compiler/z/codegen/OMRCodeGenerator.cpp#L2568-L2610

[2] https://github.com/eclipse/omr/blob/fcfd166b487d3796b525be3def5afe8b97c2c022/compiler/z/codegen/SystemLinkage.cpp#L2327-L2336

[3] https://github.com/eclipse/omr/blob/fcfd166b487d3796b525be3def5afe8b97c2c022/compiler/z/codegen/SystemLinkage.cpp#L90-L121

Reply:
OK, that's a little different than what I assumed, but still salvageable. Is the number of locked regs known and fixed by the time we do register allocation? That's kind of what it looks like based on the above.

If so, we can/should still iterate through the reg file to count locked regs; what I was suggesting was that it doesn't need to be done every time we assign regs in a set of reg deps as it is now, it can simply be done once before RA. Based on what Z is doing it might be best to do it in common OMR code immediately before RA starts, that way we don't have to do it in every linkage in OMR and downstream and can accommodate cases where the number of locked regs is determined later (after the linkage initializes the register file) as well.

As long as all code generators agree on the meaning of `Locked`, that it should be for registers that don't participate in register allocation and that registers don't change from locked to some other state or vice versa during RA it's all good.

Reply:
> what I was suggesting was that it doesn't need to be done every time we assign regs in a set of reg deps as it is now, it can simply be done once before RA

@ymanton this is exactly what happens with the above change. The code modified is in `OMR::Z::CodeGenerator::prepareRegistersForAssignment` which only gets called once per compilation before we do RA:

https://github.com/eclipse/omr/blob/fcfd166b487d3796b525be3def5afe8b97c2c022/compiler/codegen/OMRCodeGenPhase.cpp#L338-L352

Reply:
Yes, but it it's still in a Z-codegen-specific place, while on Power a hardcoded count is still in all of the linkages (both OMR and downstream) -- what I'm getting at (perhaps poorly) is that we can simply do the counting in a common pre-RA place, like ` OMR::CodeGenPhase::performRegisterAssigningPhase()`. It will introduce common behaviour on all platforms regardless of whether or not the count is a constant property of the linkage (like on Power) or whether it's somewhat more dynamic (like on Z), make it not possible for someone writing a new codegen or new linkage to forget to set it, etc. Does that make sense or am I misunderstanding?

Reply:
(I see a bit of where my own confusion came from, once upon a time on Power the locked regs used to be counted up in `TR_PPCRegisterDependencyGroup::assignRegisters()` but at some point we moved it into the linkage and got rid of the loop and left that bit of paranoid code that's now guarded by `#DEBUG`. For some reason I imagined that you somehow introduced that to Z. My mistake, sorry!)